core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,503 @@
+    assert(tracker.nextExpiryTime === Long.MaxValue)
+  }
+
+  test("multiple attempts for the same task count once") {
+    // make sure that for blacklisting tasks, the node counts task attempts, not executors.  But for
+    // stage-level blacklisting, we count unique tasks.  The reason for this difference is, with
+    // task-attempt blacklisting, we want to make it easy to configure so that you ensure a node
+    // is blacklisted before the taskset is completely aborted b/c of spark.task.maxFailures.
+    // But with stage-blacklisting, we want to make sure we're not just counting one bad task
+    // that has failed many times.
+
+    val (tracker, scheduler) = trackerFixture(
+      config.MAX_TASK_ATTEMPTS_PER_EXECUTOR.key -> "2",
+      config.MAX_TASK_ATTEMPTS_PER_NODE.key -> "3",
+      config.MAX_FAILURES_PER_EXEC_STAGE.key -> "2",
+      config.MAX_FAILED_EXEC_PER_NODE_STAGE.key -> "3"
+    )
+    val taskSet = FakeTask.createTaskSet(5)
+    val tsm = new TaskSetManager(scheduler, Some(tracker), taskSet, 4, clock)
+    // fail a task twice on hostA, exec:1
+    tsm.updateBlacklistForFailedTask("hostA", "1", 0)
+    tsm.updateBlacklistForFailedTask("hostA", "1", 0)
+    assert(tsm.isExecutorBlacklistedForTask("1", 0))
+    assert(!tsm.isNodeBlacklistedForTask("hostA", 0))
+    assert(!tsm.isExecutorBlacklistedForTaskSet("1"))
+    assert(!tsm.isNodeBlacklistedForTaskSet("hostA"))
+
+    // fail the same task once more on hostA, exec:2
+    tsm.updateBlacklistForFailedTask("hostA", "2", 0)
+    assert(tsm.isNodeBlacklistedForTask("hostA", 0))
+    assert(!tsm.isExecutorBlacklistedForTaskSet("2"))
+    assert(!tsm.isNodeBlacklistedForTaskSet("hostA"))
+
+    // fail another task on hostA, exec:1.  Now that executor has failures on two different tasks,
+    // so its blacklisted
+    tsm.updateBlacklistForFailedTask("hostA", "1", 1)
+    tracker.updateBlacklistForSuccessfulTaskSet(0, 0, tsm.execToFailures)
+    assert(tsm.isExecutorBlacklistedForTaskSet("1"))
+    assert(!tsm.isNodeBlacklistedForTaskSet("hostA"))
+
+    // fail a third task on hostA, exec:2, so that exec is blacklisted for the whole task set
+    tsm.updateBlacklistForFailedTask("hostA", "2", 2)
+    tracker.updateBlacklistForSuccessfulTaskSet(0, 0, tsm.execToFailures)
+    assert(tsm.isExecutorBlacklistedForTaskSet("2"))
+    assert(!tsm.isNodeBlacklistedForTaskSet("hostA"))
+
+    // fail a fourth & fifth task on hostA, exec:3.  Now we've got three executors that are
+    // blacklisted for the taskset, so blacklist the whole node.
+    tsm.updateBlacklistForFailedTask("hostA", "3", 3)
+    tsm.updateBlacklistForFailedTask("hostA", "3", 4)
+    tracker.updateBlacklistForSuccessfulTaskSet(0, 0, tsm.execToFailures)
+    assert(tsm.isExecutorBlacklistedForTaskSet("3"))
+    assert(tsm.isNodeBlacklistedForTaskSet("hostA"))
+  }
+
+  test("only blacklist nodes when all the blacklisted executors are all on same host (app level)") {
```

Comment:
maybe just use the more verbose "only blacklist nodes for the application when all..." and below "only blacklist nodes for the task set when all..."? took me a minute to figure out what your abbrevs meant